### PR TITLE
KAFKA-16237: Add Scala 3 support for Kafka Streams

### PR DESCRIPTION
**Pull request by mberndt123**
A port of Kafka-Streams-Scala.

The code itself was almost entirely compatible, most of the work revolves around the build system.

Because core doesn't yet support Scala 3, it's necessary to specify the Scala version for the two separately, so there is now a `streamsScalaVersion` in addition to `scalaVersion`.

The two versions need to be compatible though because the Kafka-Streams-Scala tests have a dependency on the Core project. Fortunately Scala 2.13 and Scala 3 can coexist on the classpath, so the tests will run even when compiling with different Scala versions. 

The Scala compiler options for Scala 3 are largely incompatible with the Scala 2.13 ones, so I've set them to `[]` for now as I don't know what the preferences of the Kafka developers are regarding that.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -2355,7 +2366,7 @@ project(':streams:streams-scala') {
     spotless {
       scala {
         target '**/*.scala'
-        scalafmt("$versions.scalafmt").configFile('../../checkstyle/.scalafmt.conf').scalaMajorVersion(versions.baseScala)
+        scalafmt("$versions.scalafmt").configFile('../../checkstyle/.scalafmt.conf').scalaMajorVersion('2.13')
```
Comment (jlprat) [inline build.gradle]: What's the motivation for this change? This means that we would run scalafmt with 2.13 rules regardless of the Scala version we compile. I think we could use something like https://scalameta.org/scalafmt/docs/configuration.html#scala-2-with--xsource3.

File: gradle.properties [hunk 2]
```diff
@@ -31,3 +31,5 @@ swaggerVersion=2.2.8
 task=build
 org.gradle.jvmargs=-Xmx2g -Xss4m -XX:+UseParallelGC
 org.gradle.parallel=true
+
+streamsScalaVersion=3.3.1
```
Comment (jlprat) [inline gradle.properties]: This shouldn't be here I believe.

File: gradle/dependencies.gradle [hunk 3]
```diff
@@ -41,6 +42,20 @@ if (hasProperty('scalaVersion')) {
...
+    versions["streamsScala"] = defaultScala212Version
+  } else if (streamsScalaVersion == '2.13') {
+    versions["streamsScala"] = defaultScala213Version
+  } else if (streamsScalaVersion == '3') {
+    versions["streamsScala"] = defaultScala3Version
+  } else {
+    versions["streamsScala"] = streamsScalaVersion
+  }
+} else {
+  versions["streamsScala"] = defaultScala212Version
```
Comment (jlprat) [inline gradle/dependencies.gradle]: This should be something like 
```suggestion
  versions["streamsScala"] = versions["scala"]
```
If you fail to specify `streamsScalaVersion` you should have the same version as `scalaVersion`.

**Review by jlprat — COMMENTED** [review]
Thanks @mberndt123 for your PR

`gradlewAll` needs to be modified as well. But as I mentioned in the mailing list, I think it would be beneficial to start a "DISCUSS" thread with this proposal, as there are a couple of things we might want to align.

File: build.gradle (hunk 1 above)
Comment (mberndt123) [reply build.gradle]: It's a leftover from an old version.
Originally I using the same `scalaVersion` for both Kafka-Streams-Scala and Core, and there is no Scala 3 artifact for scalafmt yet, so it wasn't working.
But since now I'm using separate versions, I'll undo this change.

**Comment by mberndt123** [comment]
> Thanks @mberndt123 for your PR
> 
> `gradlewAll` needs to be modified as well. But as I mentioned in the mailing list, I think it would be beneficial to start a "DISCUSS" thread with this proposal, as there are a couple of things we might want to align.

The only reasonable change would be to add ` && ./gradlew "$@" -PscalaVersion=2.13 -PstreamsScalaVersion=3` to the script. But then when you run something like `./gradlewAll :core:publish`, it's going to run that task twice for Scala 2.13, which is probably a bad idea. So I wouldn't know how to change that script in a useful way 🤷🏻‍♂️

File: gradle.properties (hunk 2 above)
Comment (mberndt123) [reply gradle.properties]: OK, I've removed it

**Comment by altrack** [comment]
@mberndt123 one of the possible solutions may be to attach the Scala 3 related step as a piggyback to Scala 2.13, it will be skipped when `scalaVersion=2.12`. This until we have a full Scala 3 release.
It can be done by adding an additional project called `:streams:streams-scala3` which will be dependent on `:streams:streams-scala` and can be excluded by `-x :streams:streams-scala3`. 
No change in gradlewAll or Jenkinsfile required.
So, by default and with scalaVersion=2.13 it will provide us with 2 artifacts for 2.13 and 3.
Also, in this case I don't see a benefit of having Scala 3 versions, as longs as those as LTS it should just work.

**Comment by dychko** [comment]
Any chance of getting this merged? 
We also want to use kafka streams with Scala 3, so would appreciate it!

**Comment by mjsax** [comment]
@mberndt123 --- Can you create a new PR for this, branched of `3.8` branch and open against `3.8`? -- We should try to make this work for `3.8` (if we cannot make it work on time for `3.8` -- we should not delay the release, let's aim for `3.9` which will follow shortly).

I talke to @jlprat and he wants to help on this, too.

**Comment by mberndt123** [comment]
We're just using the Scala 2.13 artifact at this point, which works fine since I removed the unnecessary dependency on scala-collection-compat. Since I can't see myself working on this Jenkins and Gradle stuff I'm closing this PR now since it's unlikely to go anywhere in the foreseeable future.

**Comment by sideeffffect** [comment]
@mberndt123 so it's possible to use the 2.13 Kafka Streams JAR from Scala 3?
